packages/@aws-cdk/aws-apigateway/lib/resource.ts

```diff
@@ -144,6 +177,130 @@ export abstract class ResourceBase extends ResourceConstruct implements IResourc
+
+    //
+    // Access-Control-Allow-Headers
+
+    const allowHeaders = options.allowHeaders || Cors.DEFAULT_HEADERS;
+    headers['Access-Control-Allow-Headers'] = `'${allowHeaders.join(',')}'`;
+
+    //
+    // Access-Control-Allow-Origin
+
+    if (options.allowOrigins.length === 0) {
+      throw new Error('allowOrigins must contain at least one origin');
+    }
+
+    // we use the first origin here and if there are more origins in the list, we
+    // will match against them in the response velocity template
+    headers['Access-Control-Allow-Origin'] = `'${options.allowOrigins[0]}'`;
+
+    //
+    // Access-Control-Allow-Methods
+
+    let allowMethods = options.allowMethods || Cors.ALL_METHODS;
+
+    if (allowMethods.includes('ANY')) {
+      if (allowMethods.length > 1) {
+        throw new Error(`ANY cannot be used with any other method. Received: ${allowMethods.join(',')}`);
+      }
+
+      allowMethods = Cors.ALL_METHODS;
+    }
+
+    headers['Access-Control-Allow-Methods'] = `'${allowMethods.join(',')}'`;
+
+    //
+    // Access-Control-Allow-Credentials
+
+    if (options.allowCredentials) {
+      headers['Access-Control-Allow-Credentials'] = `'true'`;
+    }
+
+    //
+    // Access-Control-Max-Age
+
+    let maxAgeSeconds;
+
+    if (options.maxAge && options.disableCache) {
+      throw new Error(`The options "maxAge" and "disableCache" are mutually exclusive`);
+    }
+
+    if (options.maxAge) {
+      maxAgeSeconds = options.maxAge.toSeconds();
+    }
+
+    if (options.disableCache) {
+      maxAgeSeconds = -1;
+    }
+
+    if (maxAgeSeconds) {
+      headers['Access-Control-Max-Age'] = `'${maxAgeSeconds}'`;
+    }
+
+    //
+    // Access-Control-Expose-Headers
+    //
+
+    if (options.exposeHeaders) {
+      headers['Access-Control-Expose-Headers'] = `'${options.exposeHeaders.join(',')}'`;
+    }
+
+    //
+    // statusCode
+
+    const statusCode = options.statusCode !== undefined ? options.statusCode : 204;
+
+    //
+    // prepare responseParams
+
+    const integrationResponseParams: { [p: string]: string } = { };
+    const methodReponseParams: { [p: string]: boolean } = { };
+
+    for (const [ name, value ] of Object.entries(headers)) {
+      const key = `method.response.header.${name}`;
+      integrationResponseParams[key] = value;
+      methodReponseParams[key] = true;
+    }
+
+    return this.addMethod('OPTIONS', new MockIntegration({
+      requestTemplates: { 'application/json': '{ statusCode: 200 }' },
+      integrationResponses: [
+        { statusCode: `${statusCode}`, responseParameters: integrationResponseParams, responseTemplates: renderResponseTemplate() }
+      ],
+    }), {
+      methodResponses: [
+        { statusCode: `${statusCode}`, responseParameters: methodReponseParams }
+      ]
+    });
+
+    // renders the response template to match all possible origins (if we have more than one)
+    function renderResponseTemplate() {
```

Comment:
When you do this, you also need to set the `Vary` header to `Origin` - so intermediate proxies include the request' `Origin` in their own cache key.

Reply:
Done